File: ts-plugins/tests/plugin-integration.test.ts
import { describe, it, expect } from 'vitest';
import { Tramli, InMemoryFlowStore, flowKey, MermaidGenerator } from '@unlaxer/tramli';
import type { StateProcessor, TransitionGuard, StateConfig, GuardOutput, FlowKey } from '@unlaxer/tramli';
import type { FlowContext } from '@unlaxer/tramli';
import {
  PluginRegistry, PluginReport,
  // Audit
  AuditStorePlugin, AuditingFlowStore,
  // EventStore
  EventLogStorePlugin, EventLogStoreDecorator,
  ReplayService, ProjectionReplayService, CompensationService,
  // Observability
  ObservabilityEnginePlugin, InMemoryTelemetrySink,
  // Resume
  RichResumeExecutor, RichResumeRuntimePlugin,
  // Idempotency
  InMemoryIdempotencyRegistry, IdempotencyRuntimePlugin, IdempotentRichResumeExecutor,
  // Diagram
  DiagramPlugin, DiagramGenerationPlugin,
  // Docs
  DocumentationPlugin, FlowDocumentationPlugin,
  // Lint
  PolicyLintPlugin,
  // Testing
  ScenarioTestPlugin, ScenarioGenerationPlugin,
  // Hierarchy
  EntryExitCompiler, HierarchyCodeGenerator, HierarchyGenerationPlugin,
  stateSpec, flowSpec, transitionSpec,
  // SubFlow
  GuaranteedSubflowValidator,
} from '../src/index.js';

// ─── Shared test flow ─────────────────────────────

type S = 'CREATED' | 'PENDING' | 'CONFIRMED' | 'DONE' | 'ERROR';

const config: Record<S, StateConfig> = {
  CREATED:   { terminal: false, initial: true },
  PENDING:   { terminal: false },
  CONFIRMED: { terminal: false },
  DONE:      { terminal: true },
  ERROR:     { terminal: true },
};

interface Input { value: string }
interface Middle { processed: boolean }
interface Output { result: string }

const InputKey = flowKey<Input>('Input');
const MiddleKey = flowKey<Middle>('Middle');
const OutputKey = flowKey<Output>('Output');

const proc1: StateProcessor<S> = {
  name: 'Proc1',
  requires: [InputKey],
  produces: [MiddleKey],
  process(ctx: FlowContext) {
    const input = ctx.get(InputKey);
    ctx.put(MiddleKey, { processed: true });
  },
};

const proc2: StateProcessor<S> = {
  name: 'Proc2',
  requires: [MiddleKey],
  produces: [OutputKey],
  process(ctx: FlowContext) {
    ctx.put(OutputKey, { result: 'done' });
  },
};

function testGuard(accept: boolean): TransitionGuard<S> {
  return {
    name: 'TestGuard',
    requires: [MiddleKey],
    produces: [],
    maxRetries: 3,
    validate(_ctx: FlowContext): GuardOutput {
      return accept
        ? { type: 'accepted' }
        : { type: 'rejected', reason: 'declined' };
    },
  };
}

function buildDef(accept = true) {
  return Tramli.define<S>('test', config)
    .setTtl(5 * 60 * 1000)
    .initiallyAvailable(InputKey)
    .from('CREATED').auto('PENDING', proc1)
    .from('PENDING').external('CONFIRMED', testGuard(accept))
    .from('CONFIRMED').auto('DONE', proc2)
    .onAnyError('ERROR')
    .build();
}

// ─── Tests ────────────────────────────────────────

describe('Plugin Integration', () => {

  it('plugin registry lifecycle', () => {
    const registry = new PluginRegistry<S>();
    const sink = new InMemoryTelemetrySink();
    registry
      .register(PolicyLintPlugin.defaults<S>())
      .register(new AuditStorePlugin())
      .register(new EventLogStorePlugin())
      .register(new ObservabilityEnginePlugin(sink));

    const def = buildDef();
    const report = registry.analyzeAll(def);
    expect(report.findings().length).toBeGreaterThanOrEqual(0);

    const store = new InMemoryFlowStore();
    const wrapped = registry.applyStorePlugins(store);
    expect(wrapped).toBeDefined();

    const engine = Tramli.engine(wrapped);
    registry.installEnginePlugins(engine);
  });

  it('store plugin wrapping', () => {
    const store = new InMemoryFlowStore();
    const audit = new AuditStorePlugin();
    const wrapped = audit.wrapStore(store);
    expect(wrapped).toBeInstanceOf(AuditingFlowStore);
  });

  it('audit store captures transitions', async () => {
    const def = buildDef(true);
    const store = new InMemoryFlowStore();
    const auditStore = new AuditingFlowStore(store);
    const engine = Tramli.engine(auditStore as any);

    await engine.startFlow(def, 's1',
      new Map([[InputKey as string, { value: 'test' }]]));

    expect(auditStore.auditedTransitions.length).toBeGreaterThanOrEqual(1);
    expect(auditStore.auditedTransitions[0].from).toBe('CREATED');
    expect(auditStore.auditedTransitions[0].to).toBe('PENDING');
  });

  it('engine plugin — observability installs loggers', () => {
    const store = new InMemoryFlowStore();
    const sink = new InMemoryTelemetrySink();
    const engine = Tramli.engine(store);
    const plugin = new ObservabilityEnginePlugin(sink);

    // Verify plugin installs without error
    plugin.install(engine);
    expect(plugin.descriptor().id).toBe('observability');
    expect(plugin.kind()).toBe('ENGINE');
  });

  it('rich resume classification', async () => {
    const def = buildDef(true);
    const store = new InMemoryFlowStore();
    const engine = Tramli.engine(store);
    const executor = new RichResumeExecutor(engine);

    const flow = await engine.startFlow(def, 's1',
      new Map([[InputKey as string, { value: 'test' }]]));

    const result = await executor.resume(
      flow.id, def, new Map(), 'PENDING',
    );
    expect(result.status).toBe('TRANSITIONED');
  });

  it('eventstore replay', async () => {
    const def = buildDef(true);
    const store = new InMemoryFlowStore();
    const eventStore = new EventLogStoreDecorator(store);
    const engine = Tramli.engine(eventStore as any);

    const flow = await engine.startFlow(def, 's1',
      new Map([[InputKey as string, { value: 'test' }]]));

    const events = eventStore.eventsForFlow(flow.id);
    expect(events.length).toBeGreaterThanOrEqual(1);

    const replay = new ReplayService();
    const stateAtV1 = replay.stateAtVersion(eventStore.events(), flow.id, 1);
    expect(stateAtV1).toBe('PENDING');
  });

  it('projection replay', async () => {
    const def = buildDef(true);
    const store = new InMemoryFlowStore();
    const eventStore = new EventLogStoreDecorator(store);
    const engine = Tramli.engine(eventStore as any);

    const flow = await engine.startFlow(def, 's1',
      new Map([[InputKey as string, { value: 'test' }]]));
    await engine.resumeAndExecute(flow.id, def);

    const projReplay = new ProjectionReplayService();
    const count = projReplay.stateAtVersion(
      eventStore.events(), flow.id, 999,
      {
        initialState: () => 0,
        apply: (state, event) => state + 1,
      },
    );
    expect(count).toBeGreaterThanOrEqual(2);
  });

  it('compensation service', () => {
    const store = new InMemoryFlowStore();
    const eventStore = new EventLogStoreDecorator(store);
    const compensation = new CompensationService(
      (event, cause) => ({ action: 'ROLLBACK', metadata: { reason: cause.message } }),
      eventStore,
    );

    const event = {
      flowId: 'f1', version: 1, type: 'TRANSITION' as const,
      from: 'A', to: 'B', trigger: 'proc',
      timestamp: new Date(), stateSnapshot: '{}',
    };
    const result = compensation.compensate(event, new Error('fail'));
    expect(result).toBe(true);

    const events = eventStore.eventsForFlow('f1');
    expect(events.length).toBe(1);
    expect(events[0].type).toBe('COMPENSATION');
  });

  it('diagram generation', () => {
    const def = buildDef(true);
    const plugin = new DiagramPlugin();
    const bundle = plugin.generate(def);
    expect(bundle.mermaid).toContain('stateDiagram-v2');
    expect(bundle.dataFlowJson).toContain('Input');
    expect(bundle.markdownSummary).toContain('# test');
  });

  it('documentation generation', () => {
    const def = buildDef(true);
    const plugin = new DocumentationPlugin();
    const md = plugin.toMarkdown(def);
    expect(md).toContain('# Flow Catalog: test');
    expect(md).toContain('CREATED');
    expect(md).toContain('DONE');
    expect(md).toContain('(initial)');
    expect(md).toContain('(terminal)');
  });

  it('scenario generation', () => {
    const def = buildDef(true);
    const plugin = new ScenarioTestPlugin();
    const plan = plugin.generate(def);
    expect(plan.scenarios.length).toBeGreaterThanOrEqual(3);
    expect(plan.scenarios[0].steps[0]).toContain('given flow in');
  });

  it('lint analysis', () => {
    const def = buildDef(true);
    const lintPlugin = PolicyLintPlugin.defaults<S>();
    const report = new PluginReport();
    lintPlugin.analyze(def, report);
    // OutputKey is dead data (produced but never consumed)
    const findings = report.findings();
    const deadFinding = findings.find(f => f.message.includes('never consumed'));
    expect(deadFinding).toBeDefined();
  });

  it('idempotency duplicate suppression', async () => {
    const def = buildDef(true);
    const store = new InMemoryFlowStore();
    const engine = Tramli.engine(store);
    const registry = new InMemoryIdempotencyRegistry();
    const executor = new IdempotentRichResumeExecutor(engine, registry);

    const flow = await engine.startFlow(def, 's1',
      new Map([[InputKey as string, { value: 'test' }]]));

    const r1 = await executor.resume(flow.id, def,
      { commandId: 'cmd-1', externalData: new Map() }, 'PENDING');
    expect(r1.status).toBe('TRANSITIONED');

    const r2 = await executor.resume(flow.id, def,
      { commandId: 'cmd-1', externalData: new Map() }, 'CONFIRMED');
    expect(r2.status).toBe('ALREADY_COMPLETE');
  });

  it('hierarchy entry-exit compiler', () => {
    const spec = flowSpec('OrderHierarchy', 'OrderHState');
    const parent = stateSpec('PROCESSING', { initial: true });
    parent.entryProduces.push('AuditLog');
    parent.exitProduces.push('CleanupLog');
    const child = stateSpec('VALIDATING');
    parent.children.push(child);
    spec.rootStates.push(parent);
    spec.transitions.push(transitionSpec('PROCESSING', 'DONE', 'complete'));

    const compiler = new EntryExitCompiler();
    const synth = compiler.synthesize(spec);
    expect(synth.length).toBeGreaterThanOrEqual(2);
    expect(synth.find(t => t.trigger.includes('__entry__'))).toBeDefined();
    expect(synth.find(t => t.trigger.includes('__exit__'))).toBeDefined();
  });

  it('hierarchy code generation', () => {
    const spec = flowSpec('Simple', 'SimpleState');
    spec.rootStates.push(stateSpec('A', { initial: true }));
    spec.rootStates.push(stateSpec('B', { terminal: true }));
    spec.transitions.push(transitionSpec('A', 'B', 'go'));

    const gen = new HierarchyCodeGenerator();
    const stateConfigSrc = gen.generateStateConfig(spec);
    expect(stateConfigSrc).toContain('A');
    expect(stateConfigSrc).toContain('terminal: true');

    const skeleton = gen.generateBuilderSkeleton(spec);
    expect(skeleton).toContain("Tramli.define('Simple'");
    expect(skeleton).toContain('go');
  });

  it('hierarchy generation plugin', () => {
    const spec = flowSpec('TestFlow', 'TestState');
    spec.rootStates.push(stateSpec('INIT', { initial: true }));
    spec.rootStates.push(stateSpec('END', { terminal: true }));

    const plugin = new HierarchyGenerationPlugin();
    const files = plugin.generate(spec);
    expect(files.get('TestState.ts')).toBeDefined();
    expect(files.get('TestFlowGenerated.ts')).toBeDefined();
  });

  it('diagram generation plugin via registry', () => {
    const def = buildDef(true);
    const plugin = new DiagramGenerationPlugin<S>();
    expect(plugin.descriptor().id).toBe('diagram');
    const bundle = plugin.generate(def);
    expect(bundle.mermaid).toContain('stateDiagram-v2');
  });

  it('flow documentation plugin via registry', () => {
    const def = buildDef(true);
    const plugin = new FlowDocumentationPlugin<S>();
    expect(plugin.descriptor().id).toBe('docs');
    const md = plugin.generate(def);
    expect(md).toContain('Flow Catalog');
  });

  it('scenario generation plugin via registry', () => {
    const def = buildDef(true);
    const plugin = new ScenarioGenerationPlugin<S>();
    expect(plugin.descriptor().id).toBe('scenario-tests');
    const plan = plugin.generate(def);
    expect(plan.scenarios.length).toBeGreaterThan(0);
  });

  it('subflow validator', () => {
    const parentDef = buildDef(true);
    const subConfig: Record<'SUB_A' | 'SUB_B', StateConfig> = {
      SUB_A: { terminal: false, initial: true },
      SUB_B: { terminal: true },
    };
    const subDef = Tramli.define<'SUB_A' | 'SUB_B'>('sub', subConfig)
      .from('SUB_A').auto('SUB_B', {
        name: 'SubProc', requires: [], produces: [],
        process() {},
      })
      .build();

    const validator = new GuaranteedSubflowValidator();
    // Should not throw — sub has no data requirements
    expect(() => validator.validate(parentDef, 'PENDING', subDef, new Set())).not.toThrow();
  });

  it('validator semantics unchanged with plugins', () => {
    const def = buildDef(true);

    // Capture without plugins
    const warningsWithout = def.warnings;
    const mermaidWithout = MermaidGenerator.generate(def);

    // Register all plugin types
    const registry = new PluginRegistry<S>();
    registry
      .register(PolicyLintPlugin.defaults<S>())
      .register(new AuditStorePlugin())
      .register(new EventLogStorePlugin())
      .register(new ObservabilityEnginePlugin(new InMemoryTelemetrySink()));

    // Definition is immutable — plugins do not change validation
    expect(def.warnings).toEqual(warningsWithout);
    expect(MermaidGenerator.generate(def)).toBe(mermaidWithout);

    // analyzeAll returns findings but does not mutate definition
    const report = registry.analyzeAll(def);
    expect(report.findings().length).toBeGreaterThanOrEqual(0);
    expect(def.warnings).toEqual(warningsWithout);
  });

  it('runtime adapter plugin binding', async () => {
    const def = buildDef(true);
    const store = new InMemoryFlowStore();
    const engine = Tramli.engine(store);

    const registry = new PluginRegistry<S>();
    registry.register(new RichResumeRuntimePlugin());
    registry.register(new IdempotencyRuntimePlugin(new InMemoryIdempotencyRegistry()));

    const adapters = registry.bindRuntimeAdapters(engine);
    expect(adapters.has('rich-resume')).toBe(true);
    expect(adapters.has('idempotency')).toBe(true);
  });
});

File: java-plugins/src/test/java/org/unlaxer/tramli/plugins/PluginIntegrationTest.java
package org.unlaxer.tramli.plugins;

import org.junit.jupiter.api.Test;
import org.unlaxer.tramli.*;
import org.unlaxer.tramli.plugins.api.PluginRegistry;
import org.unlaxer.tramli.plugins.audit.AuditStorePlugin;
import org.unlaxer.tramli.plugins.audit.AuditingFlowStore;
import org.unlaxer.tramli.plugins.diagram.DiagramGenerationPlugin;
import org.unlaxer.tramli.plugins.docs.FlowDocumentationPlugin;
import org.unlaxer.tramli.plugins.eventstore.*;
import org.unlaxer.tramli.plugins.idempotency.*;
import org.unlaxer.tramli.plugins.lint.PolicyLintPlugin;
import org.unlaxer.tramli.plugins.observability.*;
import org.unlaxer.tramli.plugins.resume.*;
import org.unlaxer.tramli.plugins.testing.ScenarioGenerationPlugin;

import java.time.Duration;
import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class PluginIntegrationTest {

    // ─── Shared flow definition ─────────────────────────

    enum S implements FlowState {
        CREATED(false, true), PENDING(false, false), CONFIRMED(false, false),
        DONE(true, false), ERROR(true, false);
        private final boolean terminal, initial;
        S(boolean t, boolean i) { terminal = t; initial = i; }
        @Override public boolean isTerminal() { return terminal; }
        @Override public boolean isInitial() { return initial; }
    }

    record Input(String v) {}
    record Middle(String v) {}
    record Output(String v) {}

    static final StateProcessor PROC1 = new StateProcessor() {
        @Override public String name() { return "Proc1"; }
        @Override public Set<Class<?>> requires() { return Set.of(Input.class); }
        @Override public Set<Class<?>> produces() { return Set.of(Middle.class); }
        @Override public void process(FlowContext ctx) {
            ctx.put(Middle.class, new Middle("from-" + ctx.get(Input.class).v()));
        }
    };

    static final StateProcessor PROC2 = new StateProcessor() {
        @Override public String name() { return "Proc2"; }
        @Override public Set<Class<?>> requires() { return Set.of(Middle.class); }
        @Override public Set<Class<?>> produces() { return Set.of(Output.class); }
        @Override public void process(FlowContext ctx) {
            ctx.put(Output.class, new Output("done"));
        }
    };

    static final TransitionGuard GUARD = new TransitionGuard() {
        @Override public String name() { return "TestGuard"; }
        @Override public Set<Class<?>> requires() { return Set.of(Middle.class); }
        @Override public Set<Class<?>> produces() { return Set.of(); }
        @Override public int maxRetries() { return 3; }
        @Override public GuardOutput validate(FlowContext ctx) {
            return new GuardOutput.Accepted();
        }
    };

    FlowDefinition<S> buildDef() {
        return Tramli.define("test", S.class)
                .ttl(Duration.ofMinutes(5))
                .initiallyAvailable(Input.class)
                .from(S.CREATED).auto(S.PENDING, PROC1)
                .from(S.PENDING).external(S.CONFIRMED, GUARD)
                .from(S.CONFIRMED).auto(S.DONE, PROC2)
                .onAnyError(S.ERROR)
                .build();
    }

    // ─── P2: Validator semantics regression ──────────────

    @Test
    void validatorSemanticsUnchangedWithPlugins() {
        var def = buildDef();

        // Build without plugins
        var warningsWithout = def.warnings();
        var graphWithout = def.dataFlowGraph().toMermaid();

        // Build with plugins registered (but plugins don't change definition)
        var registry = new PluginRegistry()
                .register(PolicyLintPlugin.defaults())
                .register(new AuditStorePlugin())
                .register(new EventLogStorePlugin());

        // Definition is immutable — same object, same validation
        assertEquals(warningsWithout, def.warnings());
        assertEquals(graphWithout, def.dataFlowGraph().toMermaid());
    }

    // ─── P4: Plugin registry lifecycle ───────────────────

    @Test
    void pluginRegistryLifecycle() {
        var registry = new PluginRegistry()
                .register(PolicyLintPlugin.defaults())
                .register(new AuditStorePlugin())
                .register(new EventLogStorePlugin())
                .register(new ObservabilityEnginePlugin(new InMemoryTelemetrySink()));

        var def = buildDef();
        var report = registry.analyzeAll(def);
        assertNotNull(report);
        assertNotNull(report.asText());
    }

    // ─── Store plugin wrapping ───────────────────────────

    @Test
    void storePluginWrapping() {
        var registry = new PluginRegistry()
                .register(new AuditStorePlugin())
                .register(new EventLogStorePlugin());

        FlowStore store = registry.applyStorePlugins(new InMemoryFlowStore());
        // Store should be wrapped (not the original InMemoryFlowStore)
        assertNotNull(store);
        assertNotEquals(InMemoryFlowStore.class, store.getClass());
    }

    // ─── Engine plugin installation ──────────────────────

    @Test
    void enginePluginInstallation() {
        var sink = new InMemoryTelemetrySink();
        var registry = new PluginRegistry()
                .register(new ObservabilityEnginePlugin(sink));

        var def = buildDef();
        FlowStore store = new InMemoryFlowStore();
        FlowEngine engine = Tramli.engine(store);
        registry.installEnginePlugins(engine);

        // Run a flow — observability should capture events
        engine.startFlow(def, "s1", Map.of(Input.class, new Input("test")));
        assertFalse(sink.events().isEmpty());
    }

    @Test
    void observabilityAppendModeChains() {
        var def = buildDef();
        FlowStore store = new InMemoryFlowStore();
        FlowEngine engine = Tramli.engine(store);

        // Install custom logger first
        List<String> customLog = new ArrayList<>();
        engine.setTransitionLogger(t -> customLog.add(t.from() + "->" + t.to()));

        // Install observability with append=true
        var sink = new InMemoryTelemetrySink();
        new ObservabilityPlugin(sink).install(engine, true);

        engine.startFlow(def, "s1", Map.of(Input.class, new Input("test")));

        // Both should fire
        assertFalse(customLog.isEmpty(), "custom logger should have fired");
        assertFalse(sink.events().isEmpty(), "sink should have events");
        assertEquals("CREATED->PENDING", customLog.get(0));
        assertEquals("transition", sink.events().get(0).type());
    }

    @Test
    void observabilityDefaultModeReplaces() {
        var def = buildDef();
        FlowStore store = new InMemoryFlowStore();
        FlowEngine engine = Tramli.engine(store);

        List<String> customLog = new ArrayList<>();
        engine.setTransitionLogger(t -> customLog.add(t.from() + "->" + t.to()));

        // Install without append (default)
        var sink = new InMemoryTelemetrySink();
        new ObservabilityPlugin(sink).install(engine);

        engine.startFlow(def, "s1", Map.of(Input.class, new Input("test")));

        assertTrue(customLog.isEmpty(), "custom logger should be replaced");
        assertFalse(sink.events().isEmpty(), "sink should have events");
    }

    // ─── Rich resume classification ──────────────────────

    @Test
    void richResumeClassification() {
        var def = buildDef();
        FlowStore store = new InMemoryFlowStore();
        FlowEngine engine = Tramli.engine(store);

        var flow = engine.startFlow(def, "s1", Map.of(Input.class, new Input("x")));
        assertEquals(S.PENDING, flow.currentState());

        RichResumeExecutor resume = new RichResumeRuntimePlugin().bind(engine);
        var result = resume.resume(flow.id(), def, Map.of(), flow.currentState());

        assertNotNull(result);
        assertNotNull(result.status());
    }

    // ─── Eventstore replay ───────────────────────────────

    @Test
    void eventstoreReplay() {
        var registry = new PluginRegistry()
                .register(new EventLogStorePlugin());

        FlowStore store = registry.applyStorePlugins(new InMemoryFlowStore());
        FlowEngine engine = Tramli.engine(store);

        var def = buildDef();
        var flow = engine.startFlow(def, "s1", Map.of(Input.class, new Input("x")));

        // Get event log
        if (store instanceof EventLogStoreDecorator eventLog) {
            var events = eventLog.eventsForFlow(flow.id());
            assertFalse(events.isEmpty());

            // Replay
            var replayState = new ReplayService().stateAtVersion(eventLog.events(), flow.id(), 1);
            assertNotNull(replayState);
        }
    }

    // ─── Diagram generation ──────────────────────────────

    @Test
    void diagramGeneration() {
        var def = buildDef();
        var bundle = new DiagramGenerationPlugin<S>().generate(def);
        assertNotNull(bundle);
        assertTrue(bundle.mermaid().contains("stateDiagram"));
    }

    // ─── Documentation generation ────────────────────────

    @Test
    void documentationGeneration() {
        var def = buildDef();
        var doc = new FlowDocumentationPlugin<S>().generate(def);
        assertNotNull(doc);
        assertTrue(doc.contains("test")); // flow name
    }

    // ─── Scenario generation ─────────────────────────────

    @Test
    void scenarioGeneration() {
        var def = buildDef();
        var plan = new ScenarioGenerationPlugin<S>().generate(def);
        assertNotNull(plan);
        assertFalse(plan.scenarios().isEmpty());
    }

    // ─── Lint / policy analysis ──────────────────────────

    @Test
    void lintAnalysis() {
        var def = buildDef();
        var report = new org.unlaxer.tramli.plugins.api.PluginReport();
        PolicyLintPlugin.<S>defaults().analyze(def, report);
        assertNotNull(report.asText());
    }

    // ─── Idempotency ─────────────────────────────────────

    @Test
    void idempotencyDuplicateSuppression() {
        var def = buildDef();
        FlowStore store = new InMemoryFlowStore();
        FlowEngine engine = Tramli.engine(store);

        var flow = engine.startFlow(def, "s1", Map.of(Input.class, new Input("x")));

        var idempotent = new IdempotencyRuntimePlugin(new InMemoryIdempotencyRegistry()).bind(engine);
        var r1 = idempotent.resume(flow.id(), def, new CommandEnvelope("cmd-1", Map.of()), flow.currentState());
        var r2 = idempotent.resume(flow.id(), def, new CommandEnvelope("cmd-1", Map.of()), flow.currentState());

        // Second call with same commandId should be deduplicated
        assertNotNull(r1);
        assertNotNull(r2);
    }
}
